로그인, 프로필 화면 [STEP 2] claire.pp, vapor.park

hano-ios opened the pull request:
# 로그인, 프로필 화면 [STEP 2] claire.pp, vapor.park 
* 안녕하세요 @zdodev claire & vapor 입니다 
* STEP 2 - 프로필 화면을 만들며 다음과 같은 것을 활용했습니다
    * 코드 이용, 화면 구성 
    *  모달리티 활용 화면 전환 
    * StackView 

## 화면 
* 구성 코드를 이용해서 화면을 구성했습니다. 
* 로그인 버튼 클릭시 다음 화면으로 전환되고, 
* 프로필 화면을 띄울 수 있도록 했습니다. 
* 수치는 Figma 디자인 가이드를 참조했습니다.

* 화면은 `autolayout`과 `stackView`를 활용해서 구현했습니다. 
    * 상단의 `close`, `won`,` gift`,` favourite` 버튼들은 `topButtonsView`로 묶어 사용했고, 
    * 하단의 `chat`, `phone`, `story` 버튼은 `bottomButtonsView`로 묶어서 사용했습니다. 
    * `divider`와 프로필 사진 영역은 하단 요소와` constraint`를 주어 간격을 만들었습니다.



 ## 스택뷰 
* 스택뷰를 활용해서 대부분의 UI를 구성했습니다. 
* 버튼들은 `horizontal`로 구성했고, 공백이 필요할 경우 `spacing` 속성과 `alignment`의 `equalspacing`을 이용했습니다. 
* 버튼 아래에 라벨이 필요할 경우에는 `vertical stackview`로 하나의 버튼을 구성했습니다. 

## 마무리 
코드로 구성하는 것은 처음이라 많이 헤맸는데, 덕분에 많은 것을 배울 수 있었습니다. 피드백 해주시면 감사하겠습니다!!


## 질문 
* 프로필 이미지에 `round` 처리를 할 때, `CGFloat` 값을 주어야 하는데 뷰 생성 전에는 뷰의 크기를 알 수 없으므로 `Float` 값을 주기가 애매했습니다. 그래서 뷰컨트롤러에서 뷰가 생성된 후에 <code>viewWillLayoutSubviews()</code>에서 프로필 `round` 처리하는 함수를 다시 호출해서 프로필을 둥글게 만들어줬습니다. 다만 이 방식은 추후에 여러 서브뷰의 레이아웃이 로드되면서 중복으로 호출될 수 있으므로, 이외에 더 좋은 방식이 있는지 궁금합니다!

* 그리고 프로필 이미지 round처리를 하라고 했을 때, 조건이 프로필 이미지의 width 의 30%만큼 이라고 명시되어있어서, 최대한 프로필이미지의 width를 이용하려고 했습니다만…`UIImageView`타입인 `profileImage` (프로필이미지) 의 `frame.width` 를 `viewWillLayoutSubviews` 메소드에서 호출될 때, 이미지 고유의 크기인 `width`값 ( 굉장히 큰 값 )이 찍히고 있습니다. 그래서 `profileImage`의 `round`처리를 `profileImage`의 `width` * 0.3 대신, `super` 뷰의 `width` 값을 계산한 방식을 택했습니다.  
    * 여기서 궁금한 점이, `viewWillLayoutSubviews`에서는 오토레이아웃의 계산이 끝났을테고,` profileImage`의` width`, `height`이 결정됐을텐데 왜 `profileImage`의 `width`가 고유의 이미지 사이즈를 보여주는 것인지 궁금합니다.

Commit:
env: [툴] swift lint 추가

컨벤션 준수를 위한 swift lint 추가

Commit:
feat: [UI] UI 구현

스토리보드 이용 로그인 화면 UI 구현

Commit:
fix: [UI] text field placeholder 수정

텍스트가 아닌 placeholder로 수정

Commit:
feat: [추가] 버튼 색상 변경 기능

이메일 또는 전화번호에 5글자 이상 입력시 로그인 버튼 색상 변경

Commit:
chore: [수정] emailTextFieldDidChange 코드 수정

코드 간결하게 수정

Commit:
feat: [추가] 버튼 활성화/비활성화 기능

5자를 입력하지 않으면 버튼이 비활성화되도록 구현

Commit:
fix: [UI] textField 텍스트 크기 대응

textField만 글자 크기가 변하지 않던 부분 수정

Commit:
feat: [추가] 지역화 구현(en, ko)

한국어와 영어에 지역화 구현

Commit:
feat: [UI] 글자 크기 dynamic type 설정

UI의 텍스트들 automatically adjusts font 체크

Commit:
chore: [수정] 파일명, 코드 수정

Main->Login, ViewController -> LoginViewController
조금 더 각 파일이 무엇을 하는 파일인지를 명확히 함
코드를 더 간결하게 수정함

Commit:
chore: [수정] 변수 접근 private로 수정

클래스 내부 IBOutlet 변수에 외부에서 접근할 수 없도록 private로 수정

Commit:
fix: [UI] body 폰트 변경

subtitle -> body
figma 디자인 가이드에 있는 사항으로 수정

Commit:
chore: [수정] 주석 제거

가독성을 위해 주석 제거함

Commit:
chore: [수정] parameter명 수정

bool -> isActive
조금 더 의미를 명확하게 수정함

Commit:
fix: [UI] figma 가이드에 따라 수정

figma 가이드에 맞지 않는 요소들 수정

Commit:
fix: [UI] 색상 수정

figma 가이드에 맞게 색상 수정

Commit:
feat: [추가] 이미지 추가

assets에 아이콘 이미지 추가

Commit:
feat: [추가] 프로필 UI, viewcontroller 추가

프로필 화면을 띄울 스토리보드와 뷰컨트롤러 추가

Commit:
feat: [추가] 이미지 추가

이미지 추가

Commit:
fix: [수정] 에러 구문 제거

에러 구문 제거

Commit:
Merge branch 'k-vapor.park' into k-claire.pp

# Conflicts:
#	.swiftlint.yml
#	LoginProfile.xcodeproj/project.pbxproj
#	LoginProfile/Base.lproj/Main.storyboard
#	LoginProfile/ViewController.swift

Commit:
Merge branch 'k-claire.pp' of https://github.com/yagom-academy/ios-login-profile into merged

Commit:
feat: [추가] UI 구현중

구현중.. (미완성)

Commit:
feat: [UI] 버튼 구현

close, favourite, won, gift, chat, phone, story 버튼 구현

Commit:
chore: [수정] 코드 수정

divider 삭제, makeColorIn 추가

Commit:
feat: [UI] 아이콘 두께 조절, UI 디테일 조정, divider 추가

figma 화면에 맞게 아이콘 두께를 조절했고,
전체적으로 아이콘이 균일하도록 조정함

Commit:
[fix]: [UI] 상단 버튼들 1:1 보정

상단 버튼들이 찌그러진 형태로 나타난 것 수정

Commit:
chore: [분리] 뷰 분리, 폴더 정리

버튼은 button에, 뷰는 view에 파일 저장
ProfileView -> top, bottom으로 분리해서 구현

Commit:
Merge branch 'k-claire.pp' into merged

Commit:
chore: [수정] 코드 수정

label 한번에 초기화

Commit:
fix: [UI] corner radius 수치 수정

이미지 width의 30%가 되도록 수정

Commit:
chore: [UI] stack뷰 내부 수정

bottom/topButtonsView를 View로 감싼 부분 밖으로 뺌
모호한 constraints 제거

Commit:
chore: [수정] swiftlint, 네이밍

네이밍을 명확하게 수정,
swiftlint 준수하도록 수정

Commit:
fix: [UI] 배경화면 전체화면에 맞도록 수정

이미지 파일에 따라 배경화면이 전체에 맞지 않을 수 있으므로
전체 화면에 맞도록 수정

Commit:
chore: [분리] drawProfileImage 분리

자주 호출되는 makeProfileRound()만 분리

zdodev commented on LoginProfile/LoginViewController.swift:
```diff
@@ -17,6 +17,13 @@ class LoginViewController: UIViewController {
         activateButton(true)
     }
     
+    @IBAction private func onLoginButtonPressed(_ sender: Any) {
```
메서드명에서 `on`으로 시작하는 것보다는 동사형으로 시작하도록 이름을 지어주는 것은 어떨까요?
그리고 해당 메서드는 `tap` 이벤트 시 실행되는 메서드로 보입니다. `press`와 `tap`의 차이를 한번 확인해볼 수 있을까요? 메서드 이름을 지을 때 참고가 될 수 있을 것 같습니다.🤗

zdodev commented on LoginProfile/LoginViewController.swift:
```diff
@@ -17,6 +17,13 @@ class LoginViewController: UIViewController {
     }
     
+    @IBAction private func onLoginButtonPressed(_ sender: Any) {
+        let board: UIStoryboard = UIStoryboard(name: "Profile", bundle: nil)
```
하단의 뷰 컨트롤러 이름처럼 스토리보드의 풀네임으로 작성하면 더 명확하게 스토리보드라는 것을 알 수 있지 않을까 싶었습니다! `board`라는 이름도 괜찮습니다!

zdodev commented on LoginProfile/LoginViewController.swift:
```diff
@@ -17,6 +17,13 @@ class LoginViewController: UIViewController {
     
+    @IBAction private func onLoginButtonPressed(_ sender: Any) {
+        let board: UIStoryboard = UIStoryboard(name: "Profile", bundle: nil)
+        let profileViewController: UIViewController = board.instantiateViewController(withIdentifier: "ProfileViewController")
```
`Profile.storyboard` 스토리보드 파일을 별도로 만들고 뷰 컨트롤러를 생성하신 다른 이유가 있을까요? 의견이 궁금합니다!🤔

gustn3965 replied:
스토리보드를 코드로 불러오는 작업을 해보고 싶었습니다. 그리고, 뷰컨트롤러 파일을 따로 만든이유는 프로필의 레이어 코너를 둥글게 만들기 위해 뷰 라이프 사이클 메소드를 호출하기 위해서 했습니다~!

zdodev commented on LoginProfile/ProfileViewController.swift:
```diff
@@ -0,0 +1,12 @@
+class ProfileViewController: UIViewController {    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.view = ProfileView()
```
뷰 컨트롤러의 `view` 프로퍼티에 새로운 view를 생성하여 대입해주셨네요? 뷰 컨트롤러에서 UI를 구현하지 않고 새로운 view를 구현하신 다른 이유가 있을까요? 궁금합니다!🤔
질문 주신 두 가지 질문에 대해서는 위의 답변부터 시작해야 될 것 같습니다.
구현하신 `ProfileView`를 뷰 컨트롤러에 대치하여 구현했을 때와, 뷰 컨트롤러에서 UI를 구현했을 때의 차이를 고민해보면 좋을 것 같습니다.
뷰 컨트롤러의 (레이아웃) 라이프 사이클을 한 번 살펴보면 좋을 것 같아요!
이에 대해 궁금하신 점이 있다면 코멘트나 메시지 남겨주세요!😁

gustn3965 replied:
따로 ProfileView 파일을 만든이유중 ..
* 첫번째 이유는 뷰컨트롤러가 `ProfileView`를 직접 만드는 코드를 갖게 되는 것은 책임이 많아진다라고 생각했습니다. 
* 두번째 이유는 분리하게 되면 코드도 간결해지고 책임도 나누어진다는 장점이 있다라고 생각했습니다.


`구현하신 ProfileView를 뷰 컨트롤러에 대치하여 구현했을 때와, 뷰 컨트롤러에서 UI를 구현했을 때의 차이를 고민해보면 좋을 것 같습니다.`  
* 저희 둘이서 생각했을 때는 차이가 없을 것 같습니다. 왜냐하면, 뷰컨트롤러에서 UI를 직접 구현한다라고 하면, `viewDidLoad`안에서 오토레이웃 관련된 코드를 작성합니다. 그런데 지금 저희가 구현한 코드에서는 `viewDidLoad`안에서 `ProfileView`를 초기화를 진행하고 있습니다. 그때` ProfileView init`메소드안에 코드들이 실행이 될텐데,  그 코드가 오토레이아웃 계산을 하는 코드이므로, 뷰컨트롤러에서 UI를 직접 구현한다라고 했을때와 차이가 없다라고 생각이 됩니다.  그 후 뷰컨트롤러는 레이아웃을 계산하는 라이프사이클 메소드가 호출될 것 이구요.. 
* 소대는 어떻게 생각하시나요?!

zdodev replied:
먼저 해당 문제를 `UIViewController`의 `View`를 생성한 `ProfileView`로 대치하였던 부분과 `viewWillLayoutSubviews` 에서 배치를 조정하는 구문만 보고 미숙한 코멘트를 드린 것 같아 죄송합니다.🙏🙏

확인해본 결과 `viewDidLayoutSubviews` 시점에서도 profileImage의 크기가 1000 이상의 값을 출력하는 것을 보고 해당 문제가 아님을 알게 되었습니다.

뷰 컨트롤러의 레이아웃을 더 자세히 알아본 결과, 뷰 컨트롤러의 하위 뷰까지는 레이아웃이 조정되었음을 나타내지만, 그 뷰들의 하위 뷰들은 레이아웃이 조정되었는지에 대해서는 나타내지 않는다는 것을 알게 되었습니다.

따라서 스택 뷰까지는 레이아웃 배치 조정이 되었는지는 알 수 있지만, 그 하위 뷰(profileImage, nameLabel)는 레이아웃이 조정되었다는 것을 나타내지 않는다는 부분을 확인하였습니다.

그 하위 뷰의 레이아웃에 따른 조정이 필요할 경우, `viewDidLayoutSubviews` 호출 이후에 구할 수 있었고 해당 부분에서 반영해보면 될 것 같습니다.

해당 코드와 질문 덕분에 저도 많은 공부가 되었습니다. 앞으로 더욱더 좋은 코멘트를 남길 수 있도록 하겠습니다. 감사합니다.🙇

관련 내용에 대한 링크 남겨드립니다~
[viewDidLayoutSubviews](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621398-viewdidlayoutsubviews)

zdodev commented on LoginProfile/View/BottomButtonsView.swift:
```diff
@@ -0,0 +1,72 @@
+import UIKit
+class BottomButtonsView: UIStackView {
```
타입명 마지막에 `View` 대신 `StackView`를 사용해보는 것은 어떨까요? 더 명확하게 스택뷰를 구현했다는 의미가 전해질 것 같습니다. 의견이 궁금합니다.😁

zdodev commented on LoginProfile/View/ProfileView.swift:
```diff
@@ -0,0 +1,87 @@
+import UIKit
+
+class ProfileView: UIView {
+    private var backgroundImage: UIImageView = UIImageView(image: UIImage(named: "sky"))
```
프로퍼티를 다른 값으로 변경하지 않는다면 `var` 대신 `let`을 사용해보는 것은 어떨까요? 다른 프로퍼티 공통사항입니다.😁

zdodev commented on LoginProfile/View/BottomButtonsView.swift:
```diff
@@ -0,0 +1,72 @@
+        label.text = "1:1 채팅"
+        label.textColor = .white
+        return label
+    }()
```
클로저를 사용해서 프로퍼티를 초기화해보셨네요.👍
초기값을 바로 반영하신 점 좋습니다.😁

zdodev commented on LoginProfile/View/BottomButtonsView.swift:
```diff
@@ -0,0 +1,72 @@
+        return label
+    }()
+    
+    lazy var chatView: UIStackView = {
```
외부에서 접근할 필요가 없다면 `private`으로 선언하는 것은 어떨까요? 다른 프로퍼티 및 메서드 공통사항입니다.😁

zdodev commented on LoginProfile/View/ProfileView.swift:
```diff
@@ -0,0 +1,87 @@
+        self.addSubview(topButtonsView)
+        self.addSubview(bottomButtonsView)
+        self.addSubview(divider)
+        drawFullsizeBackground()
+        drawProfileImage()
+        setConstraint(view: topButtonsView, left: 16, right: 16, top: 16, bottom: nil)
+        setConstraint(view: bottomButtonsView, left: 60, right: 60, top: nil, bottom: 100)
+        setConstraint(view: divider, left: 0, right: 0, top: nil, bottom: nil)
```
메서드의 역할을 잘 나눠서 정의해주셨네요!👍

zdodev commented on LoginProfile/View/TopButtonsView.swift:
```diff
@@ -0,0 +1,43 @@
+import UIKit
+class TopButtonsView: UIStackView {
```
`View` 대신 `StackView`를 사용해보는 것은 어떨까요? 더 명확하게 스택뷰를 구현했다는 의미가 전해질 것 같습니다. 의견이 궁금합니다.😁

zdodev reviewed (changes requested):
안녕하세요. @HanEun-ji , @gustn3965 !
step2 PR 축하합니다.🎉

UI 구현을 figma 디자인에 맞게 잘 구현해주셨습니다.
코드로 작성, 오토레이아웃도 잘 적용해주셨습니다.👍

궁금증과 의견을 좀 남겨보았습니다!
궁금하신 점이 있다면 같이 의견을 나눠보면 좋겠습니다.😁

Commit:
chore: [수정] rename, var->let, private 사용

2차 PR 코멘트에 맞게 수정
naming convention에 맞지 않는 변수/클래스명 수정
let을 사용할 수 있는 변수들은 var에서 let으로 수정
외부에서 사용하지 않는 변수/함수들은 private 사용
궁금한 사항들은 comment 남겼습니다!

gustn3965 commented:
@zdodev 
안녕하세요~ claire & vapor 입니다.
소대! 이번에도 꼼꼼하게 리뷰해주셔서 감사합니다~! 🙏

코멘트 주신 부분들은 반영했고, 반영한 코멘트들은 닫았습니다.
다만 궁금한 코멘트는 닫지 않고 밑에 추가로 코멘트 달아두었습니다. 확인해주시면 감사하겠습니다 ~! 👍

zdodev reviewed (approved):
안녕하세요. @HanEun-ji , @gustn3965 !

코드 수정 반영해주신 부분 잘 개선해주셨습니다.👍
스텝 1, 2 진행해주시면서 많은 고민을 해보셨을것 같습니다.
질문에 대한 코멘트도 남겨보았습니다.
이번 프로젝트는 여기서 마무리 하도록 하겠습니다. 고생하셨습니다.👍👍
